middleware/assert_test.go

```diff
@@ -14,106 +14,149 @@ import (
-			datastoreMock := mock.NewDataStore().Mock
-			datastoreMock.GetFilterFunc = func(ctx context.Context, filterID, etag string) (*models.Filter, error) {
-				return &models.Filter{
-					Type: "flexible",
-					Dataset: &models.Dataset{
-						Version: 1,
-					},
-					ID: filterID,
-				}, nil
-			}
-
-			Convey("When an incoming request passes through the assert.FilterType middleware", func(){
-				assert := middleware.NewAssert(
-					responder.New(),
-					&mock.DatasetAPI{},
-					filterFlexAPIMock,
-					datastoreMock,
-					testToken,
-					true,
-				)
+func assertTestSetup(filterIsFlexible, datasetIsFlexible, assertIsEnabled bool, testBluePrintID string) (*apimock.DataStoreMock, *apimock.FilterFlexAPIMock, *apimock.DatasetAPIMock, *middleware.Assert) {
+	filterFlexAPIMock := setupFilterFlexAPIMock()
+	datastoreMock := setupDataStoreMock(filterIsFlexible, testBluePrintID)
+	datasetAPIMock := setupDatasetAPIMock(datasetIsFlexible)
+
+	assert := middleware.NewAssert(
+		responder.New(),
+		datasetAPIMock,
+		filterFlexAPIMock,
+		datastoreMock,
+		testToken,
+		assertIsEnabled,
+	)
+	return datastoreMock, filterFlexAPIMock, datasetAPIMock, assert
+}
+
+func TestAssertFilterType(t *testing.T) {
+	Convey("Given a healthy dp-cantabular-filter-flex-api", t, func() {
+		testBluePrintID := ""
+		datasetIsFlexible := false
+		Convey("When a filter with given ID in the datastore of type 'flexible' is found", func() {
+			filterIsFlexible := true
+			Convey("When an incoming request passes through the assert.FilterType middleware", func() {
+				assertIsEnabled := true
+				datastoreMock, filterFlexAPIMock, _, assert := assertTestSetup(filterIsFlexible, datasetIsFlexible, assertIsEnabled, testBluePrintID)
 
 				w := httptest.NewRecorder()
 
 				testID := "12345678"
-				r, err := http.NewRequest("GET", "http://localhost:1234/test/" + testID, nil)
+				r, err := http.NewRequest("GET", "http://localhost:1234/test/"+testID, nil)
 				So(err, ShouldBeNil)
 
-				mux.SetURLVars(r, map[string]string{
+				requestWithVars := mux.SetURLVars(r, map[string]string{
 					filterBlueprintID: testID,
 				})
 
 				next := http.HandlerFunc(testHandler)
 				f := assert.FilterType(next)
-				f.ServeHTTP(w, r)
+				f.ServeHTTP(w, requestWithVars)
 
-				Convey("The response should have body, status and headers as returned by dp-filter-flex-api", func(){
+				Convey("The response should have body, status and headers as returned by dp-filter-flex-api", func() {
 					So(len(datastoreMock.GetFilterCalls()), ShouldEqual, 1)
+					So(datastoreMock.GetFilterCalls()[0].FilterID, ShouldEqual, testID)
 					So(len(filterFlexAPIMock.ForwardRequestCalls()), ShouldEqual, 1)
-					So(w.Code, ShouldEqual, expectedResponse.StatusCode)
+					So(w.Code, ShouldEqual, http.StatusOK)
 					So(w.HeaderMap.Get("X-Test"), ShouldResemble, "Value")
 					So(w.HeaderMap.Get("X-Foo"), ShouldResemble, "")
 					So(w.Body.String(), ShouldResemble, testBody)
```

Comment:
You might be able to abstract some of these into shared assertions if they're consistent:

```go
func theRequestShouldBeForwarded(actual interface{}, _ ...interface{}) string {
	w := actual.(*httptest.ResponseRecorder)
	headers := w.Header()

	if headers.Get("X-Foo") != "" {
		return "Test header for X-Foo was found"
	}

	if headers.Get("X-Test") != "Value" {
		return "Test header for X-Test was found"
	}

	if w.Body.String() != testBody {
		return "Test body was found instead of test handler response"
	}

	return ""
}

// So(w, theRequestShouldBeForwarded)
```
Maybe could even move some of these string values into constants, idk.

Reply:
If you don't go this route, still worth getting rid of `HeaderMap` as it's deprecated tho.

Reply:
> If you don't go this route, still worth getting rid of HeaderMap as it's deprecated tho

This is true, but the `httptest` package still uses it.

Reply:
> If you don't go this route, still worth getting rid of HeaderMap as it's deprecated tho

This is true, but the `httptest` package still uses it.

Reply:
I will keep the code as it is, as there are different expected values to be tested.